homeassistant/components/intellifire/sensor.py

```diff
@@ -0,0 +1,130 @@
+"""Platform for sensor integration."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+
+from intellifire4py import IntellifirePollData
+
+from homeassistant.components.sensor import (
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ATTR_ATTRIBUTION,
+    DEVICE_CLASS_TEMPERATURE,
+    TEMP_CELSIUS,
+    TIME_MINUTES,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from . import IntellifireDataUpdateCoordinator
+from .const import DOMAIN
+
+
+@dataclass
+class IntellifireSensorEntityDescriptionMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable[
+        [IntellifirePollData], int
+    ]  # Although sensours could have a variety of different return values - all the ones below are only reutrning ints
+
+
+@dataclass
+class IntellifireSensorEntityDescription(
+    SensorEntityDescription, IntellifireSensorEntityDescriptionMixin
+):
+    """Describes a sensor sensor entity."""
+
+
+ATTRIBUTION = "Data provided by unpublished Intellifire API"
+
+
+INTELLIFIRE_SENSORS: tuple[IntellifireSensorEntityDescription, ...] = (
+    IntellifireSensorEntityDescription(
+        key="flame_height",
+        icon="mdi:fire-circle",
+        name="Flame Height",
+        state_class=SensorStateClass.MEASUREMENT,
+        value_fn=lambda data: data.flameheight,
+    ),
+    IntellifireSensorEntityDescription(
+        key="temperature",
+        icon="mdi:car-brake-temperature",
+        name="Temperature",
+        state_class=SensorStateClass.MEASUREMENT,
+        device_class=DEVICE_CLASS_TEMPERATURE,
+        native_unit_of_measurement=TEMP_CELSIUS,
+        value_fn=lambda data: data.temperature_c,
+    ),
+    IntellifireSensorEntityDescription(
+        key="target_temp",
+        icon="mdi:thermometer-lines",
+        name="Target Temperature",
+        state_class=SensorStateClass.MEASUREMENT,
+        device_class=DEVICE_CLASS_TEMPERATURE,
+        native_unit_of_measurement=TEMP_CELSIUS,
+        value_fn=lambda data: data.thermostat_setpoint_c,
+    ),
+    IntellifireSensorEntityDescription(
+        key="fan_speed",
+        icon="mdi:fan",
+        name="Fan Speed",
+        state_class=SensorStateClass.MEASUREMENT,
+        value_fn=lambda data: data.fanspeed,
+    ),
+    IntellifireSensorEntityDescription(
+        key="timer_remaining",
+        icon="mdi:timer-sand",
+        name="Timer Time Remaining",
+        native_unit_of_measurement=TIME_MINUTES,
+        state_class=SensorStateClass.MEASUREMENT,
+        value_fn=lambda data: round(int(data.timeremaining_s) / 60),
```

Comment:
We don't allow for entities to "count down" time. Instead, the end time should be calculated and returned as a datetime object.

Reply:
I can do that -> the actual hardware sensor presents "seconds remaining * 100" and counts down as the time changes.... I'll need to update this I guess and maybe go with something like ... `timer_end_time`

Reply:
I'm not resolving this yet :) ... I made a change but I think its not correct


<img width="313" alt="image" src="https://user-images.githubusercontent.com/6491743/151095742-82a159c3-dd02-4ed1-a227-dff02bc7b758.png">

```python
    IntellifireSensorEntityDescription(
        key="timer_end_time",
        icon="mdi:timer-sand",
        name="Timer End",
        state_class=SensorStateClass.MEASUREMENT,
        value_fn=lambda data: (
            datetime.datetime.now() + datetime.timedelta(seconds=data.timeremaining_s)
        ).strftime("%X"),
    ),
```

I'm taking current time - adding the time offset and returning a string with the `%X` formatter. 

I'm assuming I either am missing a `native_unit_of_measurement` for times or something so it correctly gets formatted into the locale. Perhaps I should be returning an epoch time or something? 

The difference here is do I use 24 hr time or am/pm 12 hr time

Reply:
I did a little playing around and got it into a `datetime` and pull out the current `hass.config.time_zone` which is maybe what you want ... but it looks REALLY bad on the GUI and I'm not sure how helpful it would be.

<img width="311" alt="image" src="https://user-images.githubusercontent.com/6491743/151103556-f9446719-2cce-43f5-becd-7cd0b1f47de2.png">

So I'd much prefer something like:

> Timer End Time: 7:33 pm

or if I was British maybe:

> Timer End Time: 21:22

Is the 24/12 hr time an option in HA?

Searching around it sounds like this is a browser setting? In any case I'm pretty sure you will be able to answer me quickly on this so I'll probably stop trying to figure it out tonight

Reply:
The current checked in code is using US style time ...